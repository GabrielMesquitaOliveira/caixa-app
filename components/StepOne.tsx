import { useProdutos } from '@/hooks/useProdutos';
import { FormData, Produto } from '@/types';
import React, { useState } from 'react';
import { Control, FieldErrors, UseFormHandleSubmit } from 'react-hook-form';
import { ActivityIndicator, Button, ScrollView, Text, TouchableOpacity, View } from 'react-native';

interface StepOneProps {
  control: Control<FormData>;
  errors: FieldErrors<FormData>;
  handleSubmit: UseFormHandleSubmit<FormData>;
  onNextStep: (data: Partial<FormData>) => void;
  onCancel: () => void;
}

const StepOne: React.FC<StepOneProps> = ({ control, errors, handleSubmit, onNextStep, onCancel }) => {
  const { produtos, loading, error } = useProdutos();
  const [produtoSelecionado, setProdutoSelecionado] = useState<Produto | null>(null);

  const handleSelecionarProduto = (produto: Produto) => {
    setProdutoSelecionado(produto);
  };

  const handleContinuar = () => {
    if (produtoSelecionado) {
      console.log('Produto selecionado:', produtoSelecionado);
      onNextStep({ produtoId: produtoSelecionado.id });
    }
  };

  if (loading) {
    return (
      <View className="flex-1 justify-center items-center">
        <ActivityIndicator size="large" color="#007AFF" />
        <Text className="mt-4 text-gray-600">Carregando modalidades...</Text>
      </View>
    );
  }

  if (error) {
    return (
      <View className="flex-1 justify-center items-center p-4">
        <Text className="text-red-600 text-center mb-4">
          Erro ao carregar modalidades: {error}
        </Text>
        <Button title="Tentar Novamente" onPress={() => window.location.reload()} />
      </View>
    );
  }

  return (
    <View className="flex-1">
      <Text className="text-xl font-bold mb-6 text-center">Escolha sua Modalidade</Text>
      
      <ScrollView className="flex-1">
        <View className="space-y-4 px-4">
          {produtos.map((produto) => (
            <TouchableOpacity
              key={produto.id}
              className={`p-4 border rounded-lg ${
                produtoSelecionado?.id === produto.id
                  ? 'border-blue-500 bg-blue-50'
                  : 'border-gray-300 bg-white'
              }`}
              onPress={() => handleSelecionarProduto(produto)}
            >
              <Text className="font-semibold text-lg">{produto.nome}</Text>
              <Text className="text-gray-600 mt-1">{produto.descricao}</Text>
              <View className="mt-2">
                <Text className="text-sm text-gray-500">
                  Taxa: {produto.taxaJurosAnual}% a.a.
                </Text>
                <Text className="text-sm text-gray-500">
                  Prazo: {produto.prazoMinimo}-{produto.prazoMaximo} meses
                </Text>
                <Text className="text-sm text-gray-500">
                  Valor: R$ {produto.valorMinimo.toLocaleString()} - R$ {produto.valorMaximo.toLocaleString()}
                </Text>
                <Text className="text-sm text-gray-500 capitalize">
                  Categoria: {produto.categoria.replace('_', ' ')}
                </Text>
              </View>
            </TouchableOpacity>
          ))}
        </View>
      </ScrollView>

      <View className="p-4 border-t border-gray-200">
        {produtoSelecionado && (
          <View className="mb-4 p-3 bg-green-50 border border-green-200 rounded-lg">
            <Text className="font-semibold text-green-800">
              Modalidade selecionada: {produtoSelecionado.nome}
            </Text>
          </View>
        )}
        
        <View className="flex-row justify-between">
          <Button title="Cancelar" onPress={onCancel} />
          <Button
            title="Continuar"
            onPress={handleContinuar}
            disabled={!produtoSelecionado}
          />
        </View>
      </View>
    </View>
  );
};

export default StepOne;
