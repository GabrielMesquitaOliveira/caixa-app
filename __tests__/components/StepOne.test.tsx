import StepOne from '@/components/StepOne';
import { render } from '@testing-library/react-native';
import React from 'react';

// Mock the useProdutos hook
jest.mock('@/hooks/useProdutos', () => ({
  useProdutos: () => ({
    produtos: [
      {
        id: '1',
        nome: 'Empréstimo Pessoal',
        taxaJurosAnual: 12.5,
        prazoMinimo: 6,
        prazoMaximo: 36,
        valorMinimo: 1000,
        valorMaximo: 50000,
        descricao: 'Empréstimo para pessoa física',
        categoria: 'pessoa_fisica'
      }
    ],
    loading: false,
    error: null
  })
}));

describe('StepOne Component', () => {
  const mockControl = {} as any;
  const mockErrors = {} as any;
  const mockHandleSubmit = jest.fn();
  const mockOnNextStep = jest.fn();
  const mockOnCancel = jest.fn();

  it('renders correctly', () => {
    const { getByText } = render(
      <StepOne 
        control={mockControl}
        errors={mockErrors}
        handleSubmit={mockHandleSubmit}
        onNextStep={mockOnNextStep}
        onCancel={mockOnCancel}
      />
    );
    
    expect(getByText('Escolha sua Modalidade')).toBeTruthy();
    expect(getByText('Empréstimo Pessoal')).toBeTruthy();
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <StepOne 
        control={mockControl}
        errors={mockErrors}
        handleSubmit={mockHandleSubmit}
        onNextStep={mockOnNextStep}
        onCancel={mockOnCancel}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
