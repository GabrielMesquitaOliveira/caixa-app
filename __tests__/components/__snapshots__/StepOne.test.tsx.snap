// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepOne Component matches snapshot 1`] = `
<View
  className="flex-1"
>
  <Text
    className="text-xl font-bold mb-6 text-center"
  >
    Escolha sua Modalidade
  </Text>
  <RCTScrollView
    className="flex-1"
  >
    <View>
      <View
        className="space-y-4 px-4"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            className="font-semibold text-lg"
          >
            Empréstimo Pessoal
          </Text>
          <Text
            className="text-gray-600 mt-1"
          >
            Empréstimo para pessoa física
          </Text>
          <View
            className="mt-2"
          >
            <Text
              className="text-sm text-gray-500"
            >
              Taxa: 
              12.5
              % a.a.
            </Text>
            <Text
              className="text-sm text-gray-500"
            >
              Prazo: 
              6
              -
              36
               meses
            </Text>
            <Text
              className="text-sm text-gray-500"
            >
              Valor: R$ 
              1,000
               - R$ 
              50,000
            </Text>
            <Text
              className="text-sm text-gray-500 capitalize"
            >
              Categoria: 
              pessoa fisica
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    className="p-4 border-t border-gray-200"
  >
    <View
      className="flex-row justify-between"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Cancelar
          </Text>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {},
              {},
            ]
          }
        >
          <Text
            disabled={true}
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Continuar
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
